src/js/experiments/snippetFeedback.js

```diff
@@ -0,0 +1,250 @@
+            this.feedbackForm = null;
+            removeClickListener();
+          }
+        };
+
+        const removeClickListener = () => {
+          document.removeEventListener('click', outsideClickListener);
+        };
+
+        document.addEventListener('click', outsideClickListener);
+      }, 0);
+    }
+  }
+
+  /**
+   * Render the char count and set it's result on the respective element.
+   * */
+  renderCharCount(charCount) {
+    let charCountLimited =
+      charCount > this.maxCharCount ? this.maxCharCount : charCount;
+    const el = `
+        <div class="charCountRow">
+          <span>${charCountLimited} / ${this.maxCharCount}</span>
+        </div>`;
+    this.currCharCountElement.innerHTML = el;
+  }
+
+  // -- Trackers --
+  // These are responsible for sending data based on user input
+  // to amplitude.
+
+  /**
+   * trackYesOrNoButton sends a trackAction to aplitude.
+   * actions are: "docs-snippet-helpful-no"  and "docs-snippet-helpful-yes"
+   * */
+  trackYesOrNoButton(yesOrNoString) {
+    window.AnalyticsClient.trackAction(
+      `docs-snippet-helpful-${yesOrNoString}`,
+      {
+        originatingSnippet: this.snippetElement.textContent,
+        location: window.location.pathname,
+        timeOfButtonClick: Date.now(),
+      },
+    );
+  }
+
+  /**
+   * `trackFormSubmission` invokes trackAction with a payload of the original
+   * snippet, the feedback of the user, and the timeOfFormSubmission.
+   * */
+  trackFormSubmission(formContent) {
+    window.AnalyticsClient.trackAction(`docs-snippet-helpful-form-submission`, {
+      originatingSnippet: this.snippetElement.textContent,
+      feedback: formContent,
+      location: window.location.pathname,
+      wasThisHelpful: this.wasThisHelpful,
+      // for diffing this against the time that the trackYesOrNoButton was clicked
+      // as well as possible the time the copy code button was clicked (TODO add that)
+      timeOfFormSubmission: Date.now(),
+    });
+  }
+
+  // -- Helpers --
+
+  /**
+   * Construct a dom element, setting classes, text content, class, onclick etc.
+   * */
+  makeElement({ kind, text, className, onClick }) {
+    let el = document.createElement(kind);
+    if (text) el.textContent = text;
+    if (className) el.classList.add(className);
+    if (onClick) el.addEventListener('click', onClick);
+
+    return el;
+  }
+}
+
+/**
+ * init queries the dom for all snipppets
+ * TODO(blocked): when that snippet's COPY CODE button is clicked... (waiting for graduation of copy code buttons)
+ *
+ * */
+function init() {
+  // add event listeners to all code snippets on page
+  window.OptimizelyClient.getVariationName({
+    experimentKey: 'dd_docs_snippet_feedback_experiment_test',
+    groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+    experimentContainer: '.hljs',
+  }).then((variation) => {
+    if (variation === 'treatment') {
+      const snippets = document.querySelectorAll('code.hljs');
```

Comment:
seems similar to what we're doing in highlight js. do we want to consolidate this into one?

Reply:
I think this may be what Nav is referring to: https://github.com/circleci/circleci-docs/blob/9022f6e1eae6196f6f714da6b97f9b569616ee91/src/js/services/highlightJSBadge.js#L31

Previously it was `const codeSnippets = [...document.getElementsByClassName('hljs')];` in `onBeforeCodeCopied`

Reply:
Hmm, I'm not sure, what do you think @nramabad ? Should we create some kind of singleton that holds some global state or whatever for common elements that come up like this?

Reply:
sounds good to me. just only thing is to ensure you're able to write mocks/tests for it.